lib/saluki-components/src/destinations/datadog_metrics/endpoint/endpoints.rs

```diff
@@ -0,0 +1,152 @@
+#[snafu(context(suffix(false)))]
+pub enum EndpointError {
+    Parse { source: url::ParseError, domain: String },
+}
+
+#[derive(Default)]
+/// Holds the mapping between domains and their api keys.
+pub struct SingleDomainResolver {
+    domain: String,
+    api_keys: Vec<String>,
+}
+
+impl SingleDomainResolver {
+    /// The domain of the resolver.
+    pub fn domain(&self) -> &String {
+        &self.domain
+    }
+
+    /// The api keys associated with the domain.
+    pub fn api_keys(&self) -> &Vec<String> {
+        &self.api_keys
+    }
+
+    fn add_api_key(&mut self, api_key: String) {
+        self.api_keys.push(api_key);
+    }
+
+    fn set_domain(&mut self, domain: String) {
+        self.domain = domain;
+    }
+}
+
+/// Converts AdditionalEndpoints into a list of SingleDomainResolvers with its
+/// destination domain and api keys.
+pub fn create_single_domain_resolvers(
+    endpoints: &AdditionalEndpoints,
+) -> Result<Vec<SingleDomainResolver>, EndpointError> {
+    let mut resolvers = Vec::new();
+    for (domain, api_keys) in &endpoints.0 {
+        let new_domain = add_adp_version_to_domain(domain.to_string())?;
+        let mut seen = HashSet::new();
+        let mut resolver = SingleDomainResolver::default();
+        for api_key in api_keys {
+            let trimmed_api_key = api_key.trim();
+            if trimmed_api_key.is_empty() || seen.contains(trimmed_api_key) {
+                continue;
+            }
+            seen.insert(trimmed_api_key);
+            resolver.add_api_key(trimmed_api_key.to_string());
+        }
+        if !resolver.api_keys.is_empty() {
+            resolver.set_domain(new_domain.trim_end_matches("/").to_string());
+            resolvers.push(resolver);
+        }
+    }
+    Ok(resolvers)
+}
+
+impl FromStr for AdditionalEndpoints {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let endpoints: HashMap<String, Vec<String>> = serde_json::from_str(s)?;
+        Ok(AdditionalEndpoints(endpoints))
+    }
+}
+
+/// Return the correct prefix for a domain.
+pub fn get_domain_prefix(app: String) -> String {
+    let app_details = saluki_metadata::get_app_details();
+    let version = app_details.version();
+    format!(
+        "{}-{}-{}-{}.agent",
+        version.major(),
+        version.minor(),
+        version.patch(),
+        app
+    )
+}
+
+/// Prefixes the domain with the ADP version.
+pub fn add_adp_version_to_domain(domain: String) -> Result<String, EndpointError> {
+    let dd_url_regexp = Regex::new(r"^app(\.mrf)?(\.[a-z]{2}\d)?\.(datad(oghq|0g)\.(com|eu)|ddog-gov\.com)$").unwrap();
+
+    let mut new_domain = domain.clone();
+
+    // Url::Parse doesnt work without this
+    if !domain.starts_with("https://") && !domain.starts_with("http://") {
+        new_domain = format!("https://{}", domain);
+    }
+
+    let d = new_domain.clone();
+    let url = Url::parse(&new_domain).context(Parse { domain })?;
```

Comment:
I think you can make this a little cleaner, and avoid the clones, by using `with_context` instead of `context`:

```suggestion
    let new_domain = if !domain.starts_with("https://") && !domain.starts_with("http://") {
        format!("https://{}", domain)
    } else {
        domain
    };
    
    let url = Url::parse(&new_domain).with_context(|_| Parse { domain.clone() })?;
```